Added Docker functionality

sgsunder opened the pull request:
Closes #169 

Mostly new files, so there shouldn't be any conflicts with other deployment methods. Only real file change is in `config.py`, which should smartly detect if it's running inside a container and pull some stuff from environment variables, and shouldn't affect anyone running it outside Docker.

The only real point of concern is ffmpeg. I elected to build a slimmed down copy inside the Dockerfile, instead of using a debian package. It takes a lot longer to build, but saves about 150MB in final image size buy cutting out useless stuff like HEVC, subtitle codecs, etc. And if you don't clear the build cache, you usually won't have to rebuild it, even when you rebuild the rest of the image.

I don't really have a strong opinion on this though, so if you think using packages is better, that can be quickly fixed.

rr- commented on server/szurubooru/config.py:
```diff
@@ -14,14 +14,40 @@ def merge(left: Dict, right: Dict) -> Dict:
+            'index': os.getenv('ESEARCH_INDEX', 'szurubooru')
+        }
+    }
 
```
Blank line needed

rr- commented on server/szurubooru/config.py:
```diff
@@ -14,14 +14,40 @@ def merge(left: Dict, right: Dict) -> Dict:
             left[key] = right[key]
     return left
 
```
Additional blank line needed

rr- commented on server/szurubooru/config.py:
```diff
@@ -1,7 +1,7 @@
 import yaml
-
+from szurubooru import errors
 
```
Additional blank line needed

rr- commented on client/nginx.conf.docker:
```diff
@@ -0,0 +1,49 @@
+
+    access_log  /dev/stdout  main;
+
+    sendfile        on;
```
The indentation for values in this file is completely arbitrary. I'd either align it in two columns, or make everything use single spaces (like `key value;`). I'm leaning towards the latter solution.

rr- commented on server/.dockerignore:
```diff
@@ -0,0 +1,10 @@
+szurubooru/tests/*
+setup.cfg
+
+# Other Unneeded Files
```
I don't think comments in this file add much value

rr- commented on server/szurubooru/config.py:
```diff
@@ -14,14 +14,40 @@ def merge(left: Dict, right: Dict) -> Dict:
+    for key in ['POSTGRES_USER',
+                'POSTGRES_PASSWORD',
+                'POSTGRES_HOST',
+                'ESEARCH_HOST']:
```
```diff
-    for key in ['POSTGRES_USER',
-                'POSTGRES_PASSWORD',
-                'POSTGRES_HOST',
-                'ESEARCH_HOST']:
+    for key in [
+            'POSTGRES_USER',
+            'POSTGRES_PASSWORD',
+            'POSTGRES_HOST',
+            'ESEARCH_HOST'
+    ]:
```

rr- commented on server/szurubooru/config.py:
```diff
@@ -14,14 +14,40 @@ def merge(left: Dict, right: Dict) -> Dict:
+                'POSTGRES_HOST',
+                'ESEARCH_HOST']:
+        if not os.getenv(key, False):
+            raise errors.ConfigError('Environment Variable %r not set' % key)
```
```diff
-            raise errors.ConfigError('Environment Variable %r not set' % key)
+            raise errors.ConfigError(f'Environment Variable {key} not set')
```

rr- commented on server/szurubooru/config.py:
```diff
@@ -14,14 +14,40 @@ def merge(left: Dict, right: Dict) -> Dict:
+        },
+        'elasticsearch': {
+            'host': os.getenv('ESEARCH_HOST'),
+            'port': os.getenv('ESEARCH_PORT', '9200'),
```
You convert `show_sql` to a bool; shouldn't this get converted to an integer?

rr- commented on server/wait-for-es:
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env python3
+
```
Unneeded blank line

rr- commented on server/wait-for-es:
```diff
@@ -0,0 +1,32 @@
+import time
+import elasticsearch
+from szurubooru import config, errors
+
```
Missing blank line

rr- commented on server/wait-for-es:
```diff
@@ -0,0 +1,32 @@
+            time.sleep(DELAY)
+            pass
+    raise errors.ThirdPartyError('Error connecting to ElasticSearch')
+
```
Missing blank line

mserajnik commented:
> I'd make docker the only deployment process supported, e.g. I'd remove config.yaml.dist in favor of config.yaml.docker (and name the latter just config.yaml.dist); remove BUILD.md etc. etc.
The motivation is that if one wants to build things manually, they can build their own pipeline by examining the docker configuration, which means we have less documentation to support.

I see Docker as a valid _option_ for deployment and it is certainly very useful for testing, but forcing it on people feels wrong to me. E.g., the way my home server is set up, there's no way I could (or rather, would) use Docker for Deployment. I honestly don't see a good reason to do this other than being able to tell people "install with Docker, if it doesn't work for you, read the Docker docs".

What do we lose by keeping the existing build process in addition to adding Docker (and maybe making Docker the default by stating so in the docs)? There's not really a huge level of extra maintenance involved. And the documentation is already there, how often does that really need more than a few additional/corrected lines?

But if you really want to do this, I strongly suggest highlighting this change in some way because this will break any automated build/update process. I'm not sure how many people are actually using szurubooru in production, but with over 100 stars on GitHub I'd assume there are at least a few.

rr- commented:
The motivation behind making deployment with Docker the only, or at least main supported build process is similar to the reason why there few comments in the source code: rather than talking about how to deploy it through documentation, it'd demonstrate that through working usable code instead. Also it's simpler and requires much fewer steps to go through. It wouldn't break deployment for other people, since they can keep using their workflows, it's just a matter of reworking documentation.

>There's not really a huge level of extra maintenance involved.

There are no tests whether the instructions are true or not, whereas Docker is CI friendly (should we want to use one some day) so it'd be immediately obvious if something was wrong.

sgsunder commented:
Made the changes. I moved `INSTALL.md` to `INSTALL-OLD.md` and added a deprecation notice at the top, and moved `DOCKER.md` to `INSTALL.md`. Hopefully that's a decent compromise.

sgsunder commented:
Side note being that Docker images will only work on Linux with x86-64 architecture. So people who deploy on BSD or Windows (I'm sure there's at least a few) or on, say, a Raspberry Pi, won't be able to use it. (If we want ARM support, we would have to make an additional set of Dockerfiles, and I don't think the audience is large enough to make it worth the effort). Sure, they can look through the Dockerfiles as an 'installation guide' of sorts but I wouldn't want to completely break non-Docker installs.

rr- commented:
>You convert show_sql to a bool; shouldn't this get converted to an integer?

I meant converting `port` to an `int`, not `show_sql` :stuck_out_tongue: 

Anyway, I think these changes are great. I'll try to test these some time this week.

sgsunder commented:
Added.

It doesn't seem to matter whether the ES port is left as a string or an int. But it'd be best for future debugging headaches for the `docker_config()` function to return the same types as the yaml parser. I think the yaml parser auto-converts number strings to ints, right?

rr- commented:
>I think the yaml parser auto-converts number strings to ints, right?
Yes

mserajnik commented on INSTALL-OLD.md:
```diff
@@ -0,0 +1,212 @@
+**This installation guide is deprecated and might be out
+of date! It is recommended that you deploy using
+[Docker](https://github.com/rr-/szurubooru/blob/master/DOCKER.md)
```
Since you've renamed `DOCKER.md` to `INSTALL.md`, this link should be adjusted accordingly.

mserajnik commented:
> It wouldn't break deployment for other people, since they can keep using their workflows, it's just a matter of reworking documentation.

Yup, I misunderstood that part. My comment was based on the (apparently wrong) assumption that building with `config.yaml` wouldn't work "out of the box" anymore and would require people to write their own build process.

Not much else to say, but you have my approval – not that you needed it. :D

> There are no tests whether the instructions are true or not, whereas Docker is CI friendly (should we want to use one some day) so it'd be immediately obvious if something was wrong.

Fair point. On that note, what do you think setting up a simple [Travis CI](https://travis-ci.com/) configuration that runs the backend tests you already have in place and checks if the builds (frontend and backend) succeed? Of course, not in this PR.

rr- commented:
>On that note, what do you think setting up a simple Travis CI configuration that runs the backend tests you already have in place and checks if the builds (frontend and backend) succeed? Of course, not in this PR.

Actually there's #171 but I told the author to wait until the docker thing gets done.

mserajnik commented:
> Actually there's #171 but I told the author to wait until the docker thing gets done.

How did I miss that? :|

sgsunder commented:
Whoops. Fixed now

sgsunder commented:
I noticed some performance issues on my server, so I scaled down the worker processes in NGINX, and enabled gzip compression. Website feels much snappier now! Also moved the mount directory so that it matches the format of the backend container.

mserajnik commented:
@sgsunder Thanks, I'll set it up on my machine next week (probably Tuesday or Wednesday) and do some testing, as discussed with @rr- in #181. I've done some Docker "homework" last week (thanks again for your guide on the other PR), so I'm confident I'll get it running. :D

mserajnik commented on INSTALL.md:
```diff
@@ -1,206 +1,65 @@
-    user@host:szuru/client$ cd ..
-    user@host:szuru$ cp config.yaml.dist config.yaml
-    user@host:szuru$ vim config.yaml
+    user@host:szuru$ cp server/config.yaml.docker config.yaml
```
You reference  `server/config.yaml.docker` instead of `server/config.yaml.dist`, but you didn't change the actual filename.

Plus, I'm not sure if copying the file from `server` to the root directory is intended or if you forgot to prepend the target filename with `server/`. If intended, why not just keep the example file in the root directory as well?

mserajnik commented on INSTALL.md:
```diff
@@ -1,206 +1,65 @@
     Pay extra attention to these fields:
 
     - data directory,
     - data URL,
```
The data URL doesn't seem to exist anymore, at least not by default. `data` seems to be the default.

mserajnik commented on INSTALL.md:
````diff
@@ -1,206 +1,65 @@
+    user@host:szuru$ docker-compose logs -f
+    # To stop
+    user@host:szuru$ docker-compose down
+    ```
````
All in all, great instructions. The comments in `docker-compose.yml.example` are also really helpful.

The only thing I'd add is the info that the frontend can be reached via `0.0.0.0:port` (the default in `docker-compose.yml` being `0.0.0.0:8080`) as people who haven't worked with Docker before might try `localhost:port` or `127.0.0.1:port` instead.

I see that the frontend is already served via nginx. I'm not that experienced with production deployment with Docker yet, but would you expose the frontend directly on `:80`/`:443` on production and adjust the configuration of the nginx running inside the container to use HTTPS, or would you still expose to `:8080` and then reverse-proxy with an nginx that's running on "bare metal" and has the HTTPS configuration?

I'm mainly curious here, but that might also be something to add (how to best go about deploying for production with HTTPS) – could also be just a link to an external guide.

sgsunder commented:
Nice catch, made the changes.

The reason that `config.yaml.dist` was moved is because the `server/` directory is used as the root for building the backend, and the builder won't be able to get the file if it's outside that directory. Moving the build context up will make the whole build process a lot messier.

When I run this on my Macbook I actually do use `http://localhost:8080/` to visit the app. Does that not work for you? It should literally just be `http://<IP address/hostname>:<whatever port you set in the compose file>/` and I didn't want to get too hand-holdy (I figured that someone who is able to open up a terminal on their machine can figure out their own local IP).

I didn't write a guide on SSL because I have very little knowledge on that (I personally run this LAN-only). Advanced users who are deploying this public-facing probably would/should make their own compose file anyway.

For reference, [this](https://github.com/sgsunder/docker-compose-nas/blob/master/szurubooru/docker-compose.yml) is how I deploy it on my own NAS.

rr- commented:
My two cents regarding SSL - I'd delegate this matter to an external nginx instance.
Motivation

- it's not like the performance is going to drop, the bottleneck will always be python (by few orders of magnitude)
- I can host many other sites alongside szurubooru and set them up in a single place

rr- commented:
Thanks for review @mserajnik , is this ready to be merged?

mserajnik commented:
@sgsunder You are right, `localhost` does indeed work. `127.0.0.1` doesn't. Weird, I had in my head that I have to use `0.0.0.0`.

Regarding SSL: I guess that's fine, I'm also leaning towards an nginx instance outside of Docker, for the same reasons @rr- has already mentioned. I was just curious because I'm not sure how I'd best handle re-generation of the certificate (in case of Let's Encrypt) etc. if I had my HTTPS-enabled nginx inside a container as well. Mounting the certificate in a volume feels a bit weird to me. Regardless, that's not really part of this PR, so I'm fine with omitting it from the installation guide.

@rr- In my opinion, ready to be merged. 👍

rr- commented:
Thanks
